cpp/perspective/src/cpp/gnode.cpp

```diff
@@ -532,19 +567,33 @@ t_gnode::_process_column<std::string>(
 }
 
 void
-t_gnode::_process() {
+t_gnode::send(t_uindex port_id, const t_data_table& fragments) {
     PSP_TRACE_SENTINEL();
-    PSP_VERBOSE_ASSERT(m_init, "touching uninited object");
-    PSP_VERBOSE_ASSERT(
-        m_mode == NODE_PROCESSING_SIMPLE_DATAFLOW, "Only simple dataflows supported currently");
-    psp_log_time(repr() + " _process.enter");
+    PSP_VERBOSE_ASSERT(m_init, "Cannot `send` to an uninited gnode.");
 
-    std::shared_ptr<t_data_table> flattened_masked = _process_table();
-    if (flattened_masked) {
-        notify_contexts(*flattened_masked);
+    if (m_input_ports.count(port_id) == 0) {
+        std::cerr << "Cannot send table to port `" << port_id << "`, which does not exist." << std::endl;
+        return;
     }
 
-    psp_log_time(repr() + " _process.noinit_path.exit");
+    std::shared_ptr<t_port>& input_port = m_input_ports[port_id];
+    input_port->send(fragments);
+}
+
+bool
+t_gnode::process(t_uindex port_id) {
+    PSP_TRACE_SENTINEL();
+    PSP_VERBOSE_ASSERT(m_init, "Cannot `process` on an uninited gnode.");
+
+    t_process_table_result result = _process_table(port_id);
+
+    if (result.m_flattened_data_table) {
+        notify_contexts(*result.m_flattened_data_table);
+    } 
+    
+    // Whether the user should be notified - False if process_table exited
+    // early, True otherwise.
+    return result.m_should_notify_userspace;
```

Comment:
We could maybe even remove the concept of no-ops and drop `m_should_notify_userspace` altogether;  instead, we can just say if `_process_table` made it past the empty-table-checks earlier in this method, we call `notify_userspace`. This may simplify some of the aggregate logic in the `context_*` classes as well since they won't need to report this information.